app/blog/posts/good-python-code.mdx

```diff
@@ -0,0 +1,333 @@
+
+1. using the `@classmethod` decorator
+2. taking time to learn about common libraries
+3. writing simpler functions
+4. being a bit lazier - earn the abstraction
+5. decouple your implementation
+
+## Use the classmethod decorator
+
+The more complex your code's logic, the more likely it is to have a strange bug. In these cases, I find the classmethod decorator to be immensely useful. I think the best example of this was from a talk by [James Powell from Don't Use This Code](http://www.dontusethiscode.com) where he uses `pandas` as an example of a library that uses classmethods well.
+
+If you've ever used the library, you'll be familiar with the following methods.
+
+- Using a csv file ? - `pandas.read_csv(csv_file)`
+- Using a pickled object ? - `pandas.read_pickle(pickle_file_path)`
+- Using a excel file? - `pandas.read_excel(file_path)`
+
+Without a classmethod, we'd have to check the file type, then see if we've got all of the necessary information to read in the file. We'd then have to make sure this logic doesn't overlap with the other files we'd like to support. With a classmethod, I don't need to deal with complex edge cases since each classmethod supports just one use-case. 
+
+
+I recently worked on writing a script that generated embeddings using different models using SentenceTransformers, OpenAI and Cohere. This was tricky because each of these models need to be used differently, even when initialising them and I finally settled on the code below.
+
+````python
+import enum
+
+
+class Provider(enum.Enum):
+    HUGGINGFACE = "HuggingFace"
+    OPENAI = "OpenAI"
+    COHERE = "Cohere"
+
+
+class EmbeddingModel:
+    def __init__(
+        self,
+        model_name: str,
+        provider: Provider,
+        max_limit: int = 20,
+        expected_iterations: int = float("inf"),
+    ):
+        self.model_name = model_name
+        self.provider = provider
+
+    @classmethod
+    def from_hf(cls, model_name: str):
+        return cls(
+            model_name,
+            provider=Provider.HUGGINGFACE,
+            max_limit=float("inf"),
+        )
+
+    @classmethod
+    def from_openai(cls, model_name: str, max_limit=20):
+        return cls(model_name, provider=Provider.OPENAI, max_limit=max_limit)
+
+    @classmethod
+    def from_cohere(cls, model_name: str):
+        return cls(model_name, provider=Provider.COHERE)
+
+````
+
+
+There are a few things which make the code above good
+
+1. **Easier To Read**: I can determine which provider I'm using when I instantiate the class - `EmbeddingModel.from_hf` makes it clear that it's the `SentenceTransformers` package that's being used
+
+2. **Lesser Redundancy**:  I only need to pass in the values that I need to for each specific model. This makes it easy to add in additional configuration parameters down the line and be confident that it won't mess up existing functionality
+
+## Learn Common Libraries
```

Comment:
```suggestion
## Learn Standard Library. 
```

Reply:
Okay, I take that back. You do mean common libraries lol

let me also add stuff like 

`collections, functools, itertools`